### ISSUE TITLE
Directly link imxvpudec to appsink issue

### ISSUE THREAD

**Issue by Bilb** [Freescale/gstreamer-imx#91]
Hi,

I am using this pipeline to play an rtp h264 stream : 

`udpsrc port=5001 name=udpsrc caps=\"application/x-rtp,pt=96,encoding-name=H264\" ! rtph264depay ! h264parse ! imxvpudec ! imxipuvideotransform ! video/x-raw,format=UYVY ! appsink  name=sink drop=true sync=false`

It's kind of working but imxipuvideotransform has some issues with video size > 1024 so I would like to remove it from my pipeline. imxvpudec output is I420, so I would like to use it directly in my app.
But when I run this pipeline:

`udpsrc port=5001 name=udpsrc caps=\"application/x-rtp,pt=96,encoding-name=H264\" ! rtph264depay ! h264parse ! imxvpudec  ! appsink  name=sink drop=true sync=false` 

with my appsink, the vpu freezes and I get these errors  : 
```
(Program:7196): GStreamer-CRITICAL **: gst_sample_get_buffer: assertion `GST_IS_SAMPLE (sample)' failed

(Program:7196): GStreamer-CRITICAL **: gst_mini_object_ref: assertion `mini_object != NULL' failed
```

I am using gstreamer imx 0.12.0, and imx-vpu 5.4.32.
Any help on that?
Cheers

**Comment by dv1** [comment]
What GStreamer version is this? And I need more information to replicate this. A small test program would be ideal.

**Comment by Bilb** [comment]
I am using gstreamer 1.4.5

**Comment by Bilb** [comment]
Here is a sample program with my makefile. 

1. If you run it directly, it should display "new sample" for each new sample (got from rtp), but doesn't do anything with it.
2. if you remove the imxipuvideotransform and the caps after it, you should see "New Sample" 4 times (in my case), and then nothing more.

You have to send something with h264 in the 5001 port. Something like this works for me:

`gst-launch-1.0  ksvideosrc ! image/jpeg,width=1280,height=720,framerate=30/1 ! jpegdec ! openh264enc bitrate=2500000 complexity=low gop-size=10 ! queue ! rtph264pay config-interval=1 ! application/x-rtp,pt=96,encoding-name=H264 ! udpsink host=ip port=5001 `

[imxnoipu_appsink.tar.gz](https://github.com/Freescale/gstreamer-imx/files/193460/imxnoipu_appsink.tar.gz)
Let me know

**Comment by Bilb** [comment]
In fact, I can reproduce this issue using the imxipuvideotransform ! video/x-raw,format=I420.
So it looks like there is something with the I420 only.

Any ideas ?

**Comment by dv1** [comment]
Hm. Strange. I think I replicated this, but I am having other problems here at the moment. I'll clean up things here, and continue.

**Comment by Bilb** [comment]
Glad to head your were able to replicate it.
Let me know if you need anything else.

**Comment by dv1** [comment]
I was unable to reproduce the assertion. However, I was able to reproduce the freeze, and also I think I know what is going on.

You need to keep in mind that the VPU uses a fixed buffer pool. This is not something that can be avoided - it is hardwired (and the firmware is closed). Your example does not unref the GstSamples. If you link imxvpudec directly to appsink, the decoded frames - which come from the fixed-size pool - end up in your application. You do not unref the samples, so these frames are not returned to the pool. Eventually, the VPU runs out of free framebuffers, and has to wait until a new one is available. But since the buffers are never unref'd, there never are new free framebuffers, and you get the freeze.

In the imxipuvideotransform ! video/x-raw,format=I420 case, imxipuvideotransform sees that both up- and downstream both support the same caps, so it switches to passthrough - it is as if imxvpudec and appsink were linked directly. Hence the same behavior as with the direct link.

In the imxipuvideotransform ! video/x-raw,format=UYVY case, imxipuvideotransform does perform a conversion (I420->UYVY). Inside, it allocates a new framebuffer for the UYVY data (using a DMA memory allocator), then performs the conversion, and once done, unrefs the input frame, which is precisely the type of frame that is *not* unref'd in your case if imxvpudec and appsink are linked directly. However, what does eventually happen is that the DMA allocator runs out of memory, and you will no longer get data from imxipuvideotransform.

In short, it seems like an issue with GstSamples not being unref'd properly.

**Comment by Bilb** [comment]
Ok, you are right, adding an unref when I link directly imxvpudec to appsink on my sample works. I thought I did not have to unref them as I was not referencing them.

Thanks for your help, I am now able to run the sample and my app without imxipuvideotransform,  using I420 directly.
